### DISCUSSION TITLE
Exact WGS-84 and ITRF implementations?

### DISCUSSION THREAD

**Discussion by trufanov-nok** [skyfielders/python-skyfield#834] [General]
Hi,
I'm able to get a sattelite pos and vel in ITRF with this code:
```
from skyfield.api import EarthSatellite, load, wgs84
from skyfield.framelib import itrs

// read TLE to line1 and line2
ts = load.timescale()
satellite = EarthSatellite(line1, line2, '', ts)
geocentric = satellite.at(satellite1.epoch)
pos, vel = geocentric.frame_xyz_and_velocity(itrs)
```

I would like to further convert pos and vel vectors to the PZ-90 coordinate system.   
It looks like this conversion depends on a kind of ITRF used in skyfield. Is it ITRF2000, ITRF2008, ITRF2014 or something like that?  
Also it seems that ITRF flavor correlates to WGS84 implementation like G730, G873, G1150, G1674... Where I can find the details about these ones used in skyfield?

**Comment by quantenschaum** [comment]
Interesting question. Maybe the following related:
I want to generate a Nautical Almanac and compute GHA (Greewich hour angle) and Dec (declination) of Sun, Moon and stars. This can be done with `earth.at(t).observe(sun).apparent().radec("date")` or `earth.at(t).observe(sun).apparent().frame_latlon(itrs)`. I would expect the latter to return correct values, but the former matches better with official almanacs I compared the results to.

How is `frame_latlon(itrs)` defined? Does it return WGS84 coordinates?

**Reply by brandon-rhodes** [reply]
Let's start with one other data point: does `.radec()` without the `'date'` parameter match what's returned from `.frame_latlon(itrs)`, or are they different?

**Comment by brandon-rhodes** [comment]
My impression is that ITRF2000, ITRF2008, and ITRF2014 are not three new reference frames, but are implementations of the single ITRF data frame that Skyfield implements. Skyfield doesn't care about the details of how, once you get out into the world, you establish geo-located stations and then measure your position relative to them; so, happily, Skyfield doesn't need to make any adjustments as you choose an implementation and measure your observation station's location.

But if you learn that my impression is wrong, please comment with more information. Thanks!

**Reply by quantenschaum** [reply]
I do not fully understand your answer and do not know enough of this topic, but AFAIU ITRF2000, ITRF2008 and ITRF2014 are [different](https://www.researchgate.net/figure/Comparison-between-ITRF2000-and-ITRF2008-at-epoch-2011_fig3_300880909), otherwise there would be no need to define them. So, when I call `frame_latlon(itrs)`in SkyField, what ITRF definition do the returned values refer to?

**Reply by brandon-rhodes** [reply]
Here's an analogy: imagine that there's a single direction called "magnetic north". But we can only know it imperfectly, through instruments. In the year 2000, someone invents the idea of having a magnetized needle floating on a cork, and declares this to be the MN2000 standard—other scientists can use the same idea in their own labs, a magnetic needle and floating cork, and call the resulting direction "magnetic north"—even though they all understand that it's an imperfect measurement.

Then someone in 2008 publishes the idea of using a thicker needle that's more precisely magnetized, and that's mounted in a glass case that protects it from breezes. They declare this to be the NM2008 standard—yet another reproducible standard that everyone can use in their lab to get an imperfect, but better, direction to call "magnetic north".

So that's the idea. When we conceive of a measurement, we then have to iterate through different actual real-world ways to measure it. And a scientist in a paper will need to state how they measured it. Are their numbers measured with respect to NM2000 or NM2008 magnetic north?

**Reply by quantenschaum** [reply]
Ok. Using this analogy I understand it like so; Using a NM2000 compass, bearings to some objects have been taken and stored in a database. Now this DB is used in some software to calculate bearings of other objects or the same objects but at a different time. The software computes a bearing and now I want to know if that calculated bearing refers to a NM2000 or a NM2008 compass. In other words which definition of north does the calculated bearing use?

**Reply by quantenschaum** [reply]
After thinking about it, it probably is more like that; The coordinate system with mag. north as reference is as it is, it's fixed. The various NM standards define ways of how to measure mag. north and thus determine how the axis of that frame are placed in the real world. So, when performing calculations on measured data like in the example above, the result is relative to mag. north but based on data that has been acquired using an NM2000 device. The calculated values are as good as the input data. If you want to switch to NM2008, you have to measure the input data with an NM2008 device. - I think, this is what you wanted to say?

**Reply by brandon-rhodes** [reply]
Yes!

If you want to make Skyfield's ITRS more accurate, by the way, make sure you load recent polar motion data:

https://rhodesmill.org/skyfield/accuracy-efficiency.html#id1

I suppose I should add a mention of the ITRS reference frame to that page of the docs. I'll go do that now.

**Comment by miccoli** [comment]
I agree with Brandon: unless IERS coordinates of points are needed with extremely high precision, the distinction between the different ITRS realizations (from ITRF88 to ITRF2020) is immaterial for the skyfield user. Transformation parameters between ITRF2020 and past ITRFs are available at <https://itrf.ign.fr/docs/solutions/itrf2020/Transfo-ITRF2020_TRFs.txt>, but are small if compared, e.g. to polar motion. 

A useful reference is IERS technical note 36, [IERS Conventions (2010)](https://www.iers.org/IERS/EN/Publications/TechnicalNotes/tn36.html-1.htm) which clearly defines the conceptual difference between a TRS (Terrestrial Reference System) and its realization, a TRF (Terrestrial Reference Frame), see section 4.1.1 and 4.1.2. Table 4.1, contains the transformation parameters from ITRF2008 to previous ITRFs at epoch 2000, while <https://itrf.ign.fr/docs/solutions/itrf2020/Transfo-ITRF2020_TRFs.txt> contains the transformation parameters of ITRF2020 at epoch 2015.
